src/protocol/boolean/generate_random_bits.rs

```diff
@@ -1,153 +1,137 @@
-    async fn generate_random_bits(self, record_id: RecordId) -> Result<Vec<Self::Share>, Error> {
-        let triples = random_bits_triples(&self, record_id);
-
-        convert_triples_to_shares(self.narrow(&Step::ConvertShares), record_id, &triples).await
-    }
+    debug_assert!(ctx.total_records().is_specified());
+    let iter = RawRandomBitIter::<F, C> {
+        ctx: ctx.clone(),
+        record_id: RecordId(0),
+        _f: PhantomData,
+    };
+    let bits = 0..(u128::BITS - F::PRIME.into().leading_zeros());
+    convert_bits(ctx, stream_iter(iter), bits)
 }
 
-#[async_trait]
-impl<F> RandomBits<F> for UpgradedSemiHonestContext<'_, F>
+/// # Errors
+/// If the conversion is unsuccessful (usually the result of communication errors).
+/// # Panics
+/// Never, but the compiler doesn't know that.
+
+// TODO : remove this hacky function and make people use the streaming version (which might be harder to use, but is cleaner)
+pub async fn one_random_bit<F, C>(
+    ctx: C,
+    record_id: RecordId,
+) -> Result<BitDecomposed<C::Share>, Error>
 where
-    F: PrimeField + ExtendableField,
+    F: PrimeField,
+    C: UpgradedContext<F>,
+    C::Share: LinearSecretSharing<F> + SecureMul<C>,
 {
-    type Share = Replicated<F>;
-
-    /// Generates a sequence of `l` random bit sharings in the target field `F`.
-    async fn generate_random_bits(self, record_id: RecordId) -> Result<Vec<Self::Share>, Error> {
-        let triples = random_bits_triples(&self, record_id);
-
-        convert_triples_to_shares(self.narrow(&Step::ConvertShares), record_id, &triples).await
-    }
-}
-
-#[async_trait]
-impl<F: PrimeField + ExtendableField> RandomBits<F> for UpgradedMaliciousContext<'_, F> {
-    type Share = MaliciousReplicated<F>;
-
-    /// Generates a sequence of `l` random bit sharings in the target field `F`.
-    async fn generate_random_bits(self, record_id: RecordId) -> Result<Vec<Self::Share>, Error> {
-        let triples = random_bits_triples::<F, _>(&self, record_id);
-
-        // Upgrade the replicated shares to malicious, in parallel,
-        let c = self.narrow(&Step::UpgradeBitTriples);
-        let ctx = &c;
-        let malicious_triples = ctx
-            .parallel_join(triples.into_iter().enumerate().map(|(i, t)| async move {
-                ctx.narrow(&BitOpStep::from(i))
-                    .upgrade_for(record_id, t)
-                    .await
-            }))
-            .await?;
-
-        convert_triples_to_shares(
-            self.narrow(&Step::ConvertShares),
-            record_id,
-            &malicious_triples,
-        )
-        .await
-    }
+    let iter = RawRandomBitIter::<F, C> {
+        ctx: ctx.clone(),
+        record_id,
+        _f: PhantomData,
+    };
+    let bits = 0..(u128::BITS - F::PRIME.into().leading_zeros());
+    Box::pin(convert_some_bits(
+        ctx,
+        // For some reason, the input stream is polled 16 times, despite this function only calling "next()" once.
```

Comment:
todo? this seems like something we may want to better understand. there might be some fake `wake` calls somewhere which may impact performance